### ISSUE TITLE
Provide key completions for TypedDict dictionaries

### ISSUE THREAD

**Issue by brettcannon** [microsoft/pylance-release#827]
E.g., for the following code:

```Python
import typing


class Spam(typing.TypedDict):
    attr: str
    other_attr: int


def func(arg: Spam) -> None:
    print(arg[])
```

I would expect to be able to get a completion from within `arg[]` as the possible set of keys are known.

**Comment by jakebailey** [comment]
I _think_ this may be a dupe of #654.

**Comment by jakebailey** [comment]
Not entirely, as that's for building the dict; this is for getting values from it. I'm surprised we don't do this, but I guess this is a string literal completion in `[]` which isn't plain `.` member access.

**Comment by brettcannon** [comment]
Yep, it's completing to a string in an index context and subtly different from #654 for the reasons you listed.

**Comment by erictraut** [comment]
This feature is already implemented, but the usability is not great currently. The problem is that the suggestion is provided only when you type the first quote character, but the client doesn't automatically trigger a suggestion request at that point. If you place the cursor between the `[` and the `]` and type a quote, then press control-space to force the completion suggestions to be displayed, you'll actually see the available TypedDict attributes appear in the completion list.

Any suggestions for how to improve the usability here?

**Comment by jakebailey** [comment]
We do declare that `[` is a character that can trigger completion: https://github.com/microsoft/pyright/blob/86221745ca1e9122e4953ed2c63515b4de374bbf/packages/pyright-internal/src/languageServerBase.ts#L960

I think what would be an improvement would be at that point to quite literally offer up the string literals as full completions, rather requiring that a user types a quote. For example, TS does something like this:

![image](https://user-images.githubusercontent.com/5341706/104385005-4b27a500-54e7-11eb-88ea-239586d6f039.png)

**Comment by jakebailey** [comment]
(The above is an example of making the object, so that's really #654, but I just mean to point at the completion itself.)

**Comment by erictraut** [comment]
Yeah, that's a good suggestion. I just checked in a change that implements this.

**Comment by jakebailey** [comment]
A gotcha is which quote character to choose; I think TS probably scans the file to figure out the predominant one to offer.

I don't see the commit yet on pyright, but I'll take a look once it's there.

**Comment by erictraut** [comment]
Pyright also chooses the predominant quote in the document.

**Comment by jakebailey** [comment]
Appears to work as expected:

![image](https://user-images.githubusercontent.com/5341706/104388409-f2a7d600-54ed-11eb-94da-d4502b9637ee.png)

Not sure how I feel about `'` being the default choice when there aren't any quotes, though. A nit.

**Comment by jakebailey** [comment]
Slightly curious which completion item kind is being used; TS has that horizontal line icon while we have the `abc` (which in my mind means "VS Code broke and is offering me word completions"). I think `abc` is `CompletionItemKind.Text` while the other is `Enum`. Can leave it to another time to compare the icons and fix that up, since we use quoted completions for other reasons as well.

**Comment by brettcannon** [comment]
Yeah, the use of `abc` as the completion type also threw me initially and made me think it was just picking up on terms in the file.

As for which quotes to use by default, best I can tell you is Black defaults to `"`/double quotes.

**Comment by jakebailey** [comment]
My preference would be to swap this to `Constant` as `Text` is VSC's default "IDK" completion icon and that might be off-putting for some people. I'll see if I can change that for today (depending on if the power outage gets me back my machine...).

Reading the quote character, the parser seems to count and then have a boolean of `predominantSingleQuoteCharacter` which defaults to true if they're equal. That might be a further improvement later to change if we believe we want to be as close to `black` as possible in quote choicing.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.1.1, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202111-13-january-2021
